### PR TITLE
Add storage and retrieval protocol notes to CODEWALK.md.

### PR DESCRIPTION
**Pull request by anorth**

### CODE REVIEW

File: CODEWALK.md [hunk 1]
```diff
@@ -302,6 +308,58 @@ A null block count indicates the absence of any blocks mined in a previous round
...
+
+A storage client ([protocol/storage/client.go](https://github.com/filecoin-project/go-filecoin/blob/master/protocol/storage/client.go)) connects directly to a miner to propose a deal, 
+using a libp2p peer id embedded in the on-chain storage miner actor data. 
+An off-chain lookup service maps peer ids to concrete addresses, in the form of multiaddr, using a [libp2p distributed hash table](https://github.com/filecoin-project/go-filecoin/blob/master/networking.md) (DHT). 
+A client also creates a payment channel so the miner can be paid over time for storing the piece. 
+The miner responds with acceptance or otherwise.
+
+When proposing a deal, a client loads the piece data into its [IPFS block service](https://github.com/ipfs/go-blockservice). 
+This advertises the availability of the data to the network, making it available to miners. 
+A miner accepting a deal pulls the data from the client (or any other host) using the IPFS block service [bitswap protocol](https://github.com/ipfs/specs/tree/master/bitswap).
```
Comment (anorth) [inline CODEWALK.md]: @whyrusleeping could you confirm or correct our understanding here?

File: CODEWALK.md [hunk 2]
```diff
@@ -302,6 +308,58 @@ A null block count indicates the absence of any blocks mined in a previous round
...
+### Retrieval
+Retrieval mining is not necessarily linked to storage mining, although in practise we expect all storage miners to also run retrieval miners. 
+Retrieval miners serve requests to fetch content, and are not much more than a big cache and some logic to find missing pieces.
+
+The retrieval protocol and implementation are not yet fully developed. 
+At present, retrieval is not charged for, and always causes unsealing of a sector.
+
+Note that due to use of the IPFS block service for transmitting pieces, a client can issue a blockservice `get` which will find pieces in the block service of the local or any other node, 
+and fetch them without invoking retrieval protocol. 
+With time, the block service will evict stale blocks, so this is not reliable.
```
Comment (anorth) [inline CODEWALK.md]: @whyrusleeping here too

File: CODEWALK.md [hunk 3]
```diff
@@ -302,6 +308,58 @@ A null block count indicates the absence of any blocks mined in a previous round
...
+using a libp2p peer id embedded in the on-chain storage miner actor data. 
+An off-chain lookup service maps peer ids to concrete addresses, in the form of multiaddr, using a [libp2p distributed hash table](https://github.com/filecoin-project/go-filecoin/blob/master/networking.md) (DHT). 
+A client also creates a payment channel so the miner can be paid over time for storing the piece. 
+The miner responds with acceptance or otherwise.
+
+When proposing a deal, a client loads the piece data into its [IPFS block service](https://github.com/ipfs/go-blockservice). 
+This advertises the availability of the data to the network, making it available to miners. 
+A miner accepting a deal pulls the data from the client (or any other host) using the IPFS block service [bitswap protocol](https://github.com/ipfs/specs/tree/master/bitswap).
+
+A miner packs pieces from many clients into a sector, which is then sealed with a proof (aka commitment).
```
Comment (phritz) [inline CODEWALK.md]: proof of replication aka commitment

File: CODEWALK.md [hunk 4]
```diff
@@ -302,6 +308,58 @@ A null block count indicates the absence of any blocks mined in a previous round
...
+encrypting, adding error-correcting codes, and replicating widely enough to achieve redundancy goals. 
+In the future, we will build a client library which handles many of these tasks for a user. 
+We also plan support for “repair miners”, to whom responsibility can be delegated for monitoring and repairing faults.
+
+### Retrieval
+Retrieval mining is not necessarily linked to storage mining, although in practise we expect all storage miners to also run retrieval miners. 
+Retrieval miners serve requests to fetch content, and are not much more than a big cache and some logic to find missing pieces.
+
+The retrieval protocol and implementation are not yet fully developed. 
+At present, retrieval is not charged for, and always causes unsealing of a sector.
```
Comment (phritz) [inline CODEWALK.md]: at present (q1'19)

**Comment by mishmosh** [comment]
@whyrusleeping please review?

File: CODEWALK.md (hunk 1 above)
Comment (whyrusleeping) [reply CODEWALK.md]: this is currently what is happening, however we will be changing this in the near future, to at least just use bitswap to pull from a single peer, depending on how graphsync work goes, we may use graphsync (bitswap v2 basically) to fetch the data.

File: CODEWALK.md (hunk 2 above)
Comment (whyrusleeping) [reply CODEWALK.md]: this is a side effect and should not be relied on. I hesitate to even include it in any documentation.

File: CODEWALK.md (hunk 1 above)
Comment (whyrusleeping) [reply CODEWALK.md]: this also depends on the resolution of this issue: https://github.com/filecoin-project/specs/issues/136

File: CODEWALK.md (hunk 2 above)
Comment (anorth) [reply CODEWALK.md]: Good call, removed.

**Pull request merged**